internal/venice-common/src/main/java/com/linkedin/venice/helix/StringAnnotatedHelixReadOnlySchemaRepository.java

```diff
@@ -0,0 +1,150 @@
+package com.linkedin.venice.helix;
+
+import static com.linkedin.venice.schema.SchemaUtils.getAnnotatedStringMapDerivedSchemaEntry;
+import static com.linkedin.venice.schema.SchemaUtils.getAnnotatedStringMapValueSchemaEntry;
+
+import com.linkedin.venice.meta.ReadOnlySchemaRepository;
+import com.linkedin.venice.schema.SchemaEntry;
+import com.linkedin.venice.schema.rmd.RmdSchemaEntry;
+import com.linkedin.venice.schema.writecompute.DerivedSchemaEntry;
+import com.linkedin.venice.utils.Pair;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+
+/**
+ * This class is a wrapper schema repository class, which is only used by merge conflict resolver in A/A store ingestion
+ * task. This class will annotate value schema and partial update schema so when performing Map collection merging,
+ * Map field's key can be deserialized into String type and thus is comparable for DCR purpose. Without this wrapper or
+ * user annotation, map fields will have key in UTF-8 type which is not comparable.
+ */
+public class StringAnnotatedHelixReadOnlySchemaRepository implements ReadOnlySchemaRepository {
+  private final ReadOnlySchemaRepository internalSchemaRepo;
+  private final Map<String, Map<Integer, SchemaEntry>> valueSchemaEntryMapCache = new VeniceConcurrentHashMap<>();
+  private final Map<String, Map<String, DerivedSchemaEntry>> partialUpdateSchemaEntryMapCache =
+      new VeniceConcurrentHashMap<>();
+
+  public StringAnnotatedHelixReadOnlySchemaRepository(ReadOnlySchemaRepository internalSchemaRepo) {
+    this.internalSchemaRepo = internalSchemaRepo;
+  }
+
+  @Override
+  public void refresh() {
+    internalSchemaRepo.refresh();
+  }
+
+  @Override
+  public void clear() {
+    internalSchemaRepo.clear();
+  }
+
+  @Override
+  public SchemaEntry getKeySchema(String storeName) {
+    return internalSchemaRepo.getKeySchema(storeName);
+  }
+
+  @Override
+  /**
+   * Retrieve value schema of a store and annotate its map fields. The annotation will only be done once in the repository's
+   * lifetime as the result is cached.
+   */
+  public SchemaEntry getValueSchema(String storeName, int id) {
+    if (!(valueSchemaEntryMapCache.containsKey(storeName) && valueSchemaEntryMapCache.get(storeName).containsKey(id))) {
```

Comment:
Can we use `computeIfAbsent` to avoid duplicate annotation calls to the same schema?